### DISCUSSION TITLE
Experimental flag?

### DISCUSSION THREAD

**Discussion by themeatsweats** [itzg/docker-minecraft-bedrock-server#373] [Q&A]
Good evening. Trying to get a couple addons working for a peaceful world and some require the Holiday Creator setting to be enabled. This is grayed out for me (even as op). I didn't see an env for it, either. Any advice on enabling this?

**Comment by itzg** [comment]
[And I don't see any mention of a server property](https://minecraft.wiki/w/Experimental_Gameplay#Toggling), so I'm not sure how this can be enabled for the server container. 

I'm going to convert this to a discussion so we can leave it around for people to answer.

**Comment by crocog1** [comment]
As per : https://old.reddit.com/user/ProfessorValko/comments/9f438p/bedrock_dedicated_server_tutorial/

> You cannot use the following types of add-ons on your server:
> Add-Ons purchased from the Minecraft Marketplace.
> Add-Ons using Experimental Gameplay features such as custom mob entities, animations, data driven spawning, particles, functions.

And the Holiday Creator stuff is experimental I think?

**Reply by josephdpurcell** [reply]
It seems there's a possibility you can enable experimental on the server, see https://github.com/perrochon/simple-bedrock-script#patch-levelsdat where it says you can edit `world/world_name/level.dat`. In my dat file I don't see that value but I do see the value documented https://minecraft.fandom.com/wiki/Bedrock_Edition_level_format which is `experiments_ever_used` and I wonder if you can set that to true if it would enable experiments.

This file is NBT format and there are some editors:

* https://nbt.mcph.to/ (online) with code repo here: https://github.com/iRath96/webNBT
* https://gamertools.net/tools/13 (online)
* https://www.minecraftforum.net/forums/mapping-and-modding-java-edition/minecraft-tools/1262665-nbtexplorer-nbt-editor-for-windows-and-mac
* https://github.com/jaquadro/NBTExplorer
* https://github.com/unascribed/unbted

There's more but these are the ones I felt comfortable enough trying. If you go searching you'll find enormous amounts of phony sites.

I couldn't get any of them to open my level.dat file. NBTExplorer segfault'd and all the others showed empty.

I don't know enough about binary editing to know if you could just edit the file with a raw editor; there are other boolean values in there.

Effectively it would seem experimental is unavailable to dedicated servers.

**Reply by zentavr** [reply]
So how to enable "Holiday Creator Features"  in this case? Which flag should be enabled?

**Reply by zentavr** [reply]
In order to to enable "Holiday Creator Features" add the key **experiments/data_driven_items** with **1** value. All the keys are here: https://wiki.bedrock.dev/nbt/experimental-education-edition.html

**Reply by josephdpurcell** [reply]
I'm still stuck trying to find a NBTEditor on Ubuntu or Mac. Here is a comment on one project with hope there is something simple I'm doing wrong https://github.com/Howaner/NBTEditor/issues/8.

**Reply by josephdpurcell** [reply]
I was able to get `jaquadro/NBTExplorer` to build! See https://github.com/jaquadro/NBTExplorer/issues/114#issuecomment-2259378041.

However, that cannot read the `level.dat` file from Bedrock.

**Reply by josephdpurcell** [reply]
I also tried https://github.com/tryashtar/nbt-studio which wasn't able to run on Linux.

Good news, however! I also was unable to edit level.dat files from bedrock using https://marketplace.visualstudio.com/items?itemName=Misodee.vscode-nbt!

**Comment by Dabbadoeber** [comment]
This program can read the file : https://github.com/tryashtar/nbt-studio

Then i added these tags : https://wiki.bedrock.dev/nbt/experimental-education-edition.html

And was abble to add mods

**Reply by josephdpurcell** [reply]
I was unable to use nbt-studio on linux, but as an alternative the VSCode plugin works https://marketplace.visualstudio.com/items?itemName=Misodee.vscode-nbt

**Comment by josephdpurcell** [comment]
To enable experimental flags you must edit the world's `level.dat` file. The following instructions are for Bedrock. See https://github.com/itzg/docker-minecraft-server/discussions/2911#discussioncomment-10209498 for confirmation that Java uses different properties.

<details>
<summary>Background...</summary>

This was a HUGE pain to figure out.

The first problem was figuring out how to edit these NBT files. You can see my journey in comments over here: https://github.com/itzg/docker-minecraft-bedrock-server/discussions/373#discussioncomment-7874376. I don't know if I saw it in a comment or if it was just on whim that I thought to try the VSCode extension. I vaguely recall seeing it before but was turned off because it said "Viewer" and I incorrectly assumed it couldn't edit files.

The second problem was figuring out what exactly needed to be edited in these files. Fortunately, @itzg's comment over here https://github.com/itzg/docker-minecraft-bedrock-server/discussions/274#discussioncomment-3553418 sparked the idea that I could use Minecraft Bedrock Launcher to create a world with the experimental flags enabled then inspect a vanilla `level.dat` file with the one that has the experiments enabled. Whoopeee!!! It worked! I was able to see exactly the flags needed.
</details>

Install a NBT editor. On Linux, the only one I could get to work was the VSCode extension [NBT Viewer](https://marketplace.visualstudio.com/items?itemName=Misodee.vscode-nbt).

Stop the server. Make a backup of the `data/` directory in case your world breaks.

Make a backup of your `level.dat` in case you need it:

```
$ cp 'data/worlds/Bedrock level/level.dat' ~/Downloads/level.dat.bak
```

Now, edit the real `level.dat` file in VSCode:

```
$ code 'data/worlds/Bedrock level/level.dat' 
```

Scroll down to "experiments" and click the + to expand the options. Double-click the `0` next to the `experiments_ever_used` and `saved_with_toggled_experiments` and replace the `0` with a `1` and then click "Confirm" like so:

![Screenshot from 2024-08-01 04-48-24](https://github.com/user-attachments/assets/66569d30-a950-40ee-b135-ec35fc4b3602)

Next we need to add flags to the "experiments" section. To do this, left click once on the word "experiments" then click "Add Tag". You should now see an input for Name and Value like so:

![Screenshot from 2024-08-01 04-49-52](https://github.com/user-attachments/assets/53f8c53c-f52f-4eba-94ca-fa0e3e02e829)

The flags you can add are all "Byte" types with a value of `1`. Here is a map from the text displayed in the UI and the key/value you'll use in the file:

* Villager Trade Rebalancing: `villager_trades_rebalance: 1`
* Holiday Creator Features: `data_driven_items: 1`
* Custom biomes: `data_driven_biomes: 1`
* Upcoming Creator Features: `upcoming_creator_features: 1`
* Beta APIs: `gametest: 1`

Save the file.

Here is an example final result:

![Screenshot from 2024-08-01 05-04-49](https://github.com/user-attachments/assets/302b3791-2d1e-4bf9-9487-e2117645c449)

Start the server. You should see the experimental flags enabled in the UI like so:

![Screenshot from 2024-08-01 05-41-48](https://github.com/user-attachments/assets/9e3d6830-89b3-41c2-bc2d-f461d87ae91e)

Further reading:

* https://old.reddit.com/r/Minecraft/comments/11l5c4k/multi_server_with_120_additions/jbbxao3/ - Java instructions for enabling experimental flags
* https://github.com/itzg/docker-minecraft-server/discussions/2911#discussioncomment-10209498 - Java instructions confirmed to work
* https://minecraft.fandom.com/wiki/Bedrock_Edition_level_format - Documentation on the `level.dat` file which entirely omits details on experimental flags
* https://feedback.minecraft.net/hc/en-us/community/posts/360048142731-Experimental-MODE-FOR-Dedicated-servers-BEDROCK - People are upset that experimental mode is not available for dedicated servers, perhaps this is referring to the fact that the experimental flags are greyed out and you have to edit the `level.dat` file manually?

**Reply by josephdpurcell** [reply]
@Cassock After much research I have found an answer! See above.

**Reply by xx1182** [reply]
Hi, there is a new update with "vibrant visuals" and it's an experiment only on beta(which I added trough the dockerfile) I do not know what string to set. Regards.

**Reply by norcino** [reply]
![image](https://github.com/user-attachments/assets/d25ca36b-4227-425d-a2c3-75f717b68d96)

Vibrant visuals is called: **experimental_graphics**
The 2025 drop 2: **y_2025_drop_2**

**Reply by brkirkpatrick** [reply]
@josephdpurcell @itzg Not sure if I should create a new topic, but I am trying to enable 2025 drop 3. I've edited the level.dat, but every time I start server it overwrites level.dat with some sort of default/existing and reverts the experimental changes. /data persists as expected, so not sure what is causing this... Any help appreciated :)

**Reply by itzg** [reply]
The only advice I can think of is to make sure the you're editing level.dat while the server is stopped.

**Reply by brkirkpatrick** [reply]
@itzg Indeed, it was stopped. Was able to make it work by copying world to single-player, enabling experiments, then copying back to server. Not sure why the steps above didn't work for me, but found another way at least ;/. Thanks!

**Reply by rolek** [reply]
Container stopped. Anyone knows why after starting container, the level.dat is reverted ?
I deleted it, deleted *.old one, uploaded new file. It keeps getting reverted. Cannot enable experiments.

**Reply by josephdpurcell** [reply]
Can you share as much detail about the steps you are taking?

I vaguely recall taking a fresh level.dat file and editing it. Consider if that works? Perhaps, for example, if it's an older level.dat it's considered invalid and recreated.

**Reply by rolek** [reply]
Exact steps:
- stop container
- delete *.old file
- backup the level.dat
- edit the level.dat using editor, set experimental flags etc.
- start container.

Log says experimental flag not enabled, level.dat file is reverterd.
